### PR TITLE
feat(scripts/ops/safeDeployRewarder.ts): self serve workflow

### PR DESCRIPTION
**Pull request by ericjuta**
- github actions checks if PR and branch with rewarder/** has opened
- runs hardhat script
- script checks if newRewarder.json file is made
- Deploys new rewarder, transfers ownership to the DAO, proposes via gnosis safe to add rewarder to MCV2
- Writes to rewarderConfigs.json file (newRewarderConfig.json with deployed rewarder address too), github actions bot auto commits that file
- Deletes newRewarderConfig.json
- Tri team can merge in

---
Adds env vars: SAFE_SIGNER_PK, SAFE_ADDRESS, AURORA_URL, SAFE_SERVICE_URL

---

NOTES -

[newRewarderConfig.sample.json](https://github.com/trisolaris-labs/trisolaris_core/blob/a952cde1b5f4d266d1b540db8adfbde91460b80a/newRewarderConfig.sample.json)

[type RewarderConfig](https://github.com/trisolaris-labs/trisolaris_core/pull/82/files#diff-dca65ff273906a16f391a5adb0a2dd6e75cde2da814f3ef118f93e9ea70be720R10)

these both need to be updated if the Config JSON schema changes.

[rewarderConfigs.json](https://github.com/trisolaris-labs/trisolaris_core/blob/a952cde1b5f4d266d1b540db8adfbde91460b80a/rewarderConfigs.json) is an empty array right now but can be populated manually (to migrate and centralise existing configs or for error handling) and automatically by the github actions bot.

rewarderConfigs.json type is `Array<RewarderConfig>`

### CODE REVIEW

**Comment by rej156** [comment]
@0xchain1 I added a // TODO: for you to combine your scripts to make them more self serve

https://github.com/trisolaris-labs/trisolaris_core/pull/82/files#diff-dca65ff273906a16f391a5adb0a2dd6e75cde2da814f3ef118f93e9ea70be720R70

File: scripts/constants.ts [hunk 1]
```diff
@@ -41,6 +41,7 @@ export const dfRecepientAddress = "0x49D8B1389df580B1602Ae385D32b9c5A7Ceb2e25";
 export const kRecepientAddress = "0x464bCBD80Ab3Ec89A93F1667DafB682d14634A5e";
 export const donDeployerAddress = "0x25b9B32d875E4b1b0ec8b74ecF4f0A2aF8e96322";
 export const specialistAddress = "0x7d479275bCa2795394e42e07a2EE2d27c96F69Ca";
+export const safeAddress = "0xf86119de6ee8d4447C8219eEC20E7561d09816d3";
```
Comment (baboobhaiya) [inline scripts/constants.ts:44]: same as this: https://github.com/trisolaris-labs/trisolaris_core/pull/82/files#diff-9a30e949794c865b1a1c4488c095866040b3e264e6ae26b1c8bf7c06b97f9178R11

File: scripts/constants.ts [hunk 2]
```diff
@@ -2,6 +2,8 @@ import { ethers } from "hardhat";
 
 export const decimals = ethers.BigNumber.from("1000000000000000000");
 export const totalSupply = ethers.BigNumber.from("1000000").mul(500).mul(decimals);
+export const safeServiceURL = "https://safe-transaction.aurora.gnosis.io/";
+export const auroraURL = "https://mainnet.aurora.dev/";
```
Comment (specialist0x) [inline scripts/constants.ts:6]: `env` var with our API key?

File: scripts/ops/safeDeployRewarder.ts [hunk 3]
```diff
@@ -0,0 +1,91 @@
...
+import { SafeEthersSigner, SafeService } from "@gnosis.pm/safe-ethers-adapters";
+import { Wallet } from "ethers";
+import { JsonRpcProvider } from "@ethersproject/providers";
+import { auroraURL, chefV2Address, safeServiceURL } from "../constants";
+
+type RewarderConfig = {
+  lpToken: string;
+  rewardToken: string;
+};
+const addNewRewarderConfigToExistingJSON = async (newRewarderConfig: RewarderConfig) => {
```
Comment (specialist0x) [inline scripts/ops/safeDeployRewarder.ts:14]: Does this need to be async?

File: scripts/ops/safeDeployRewarder.ts [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+    writeFileSync("./rewarderConfigs.json", JSON.stringify(rewarderConfigsJSON));
+    console.info("*** Added new rewarder config to rewarderConfigs.json");
+  } catch (err) {
+    console.error(err);
+  }
+};
+
+const _proposeAddingNewRewarderToSafe = async (newRewarderConfig: RewarderConfig) => {
+  const safeAddress = process.env.SAFE_ADDRESS;
+  const safeSignerPK = process.env.SAFE_SIGNER_PK;
```
Comment (specialist0x) [inline scripts/ops/safeDeployRewarder.ts:30]:
```suggestion
  const { SAFE_ADDRESS: safeAddress = undefined, SAFE_SIGNER_PK: safeSignerPK = undefined } = process.env
```

File: scripts/ops/safeDeployRewarder.ts [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+      const newRewarderConfigJSONFile = readFileSync("./newRewarderConfig.json");
+      const newRewarderConfig = JSON.parse(newRewarderConfigJSONFile?.toString());
+      console.log(newRewarderConfig);
+
+      // TODO: 0xchain to add combined scripts here
+      // Add Rewarder to MCV2 etc.
+      // await proposeAddingNewRewarderToSafe(newRewarderConfig);
+      //
+
+      await addNewRewarderConfigToExistingJSON(newRewarderConfig);
```
Comment (specialist0x) [inline scripts/ops/safeDeployRewarder.ts:75]: Should the `newRewarderConfig.json` file be deleted after `addNewRewarderConfigToExistingJSON` successfully runs?

File: scripts/ops/safeDeployRewarder.ts [hunk 6]
```diff
@@ -0,0 +1,91 @@
+import { writeFileSync, existsSync, readFileSync } from "fs";
```
Comment (specialist0x) [inline scripts/ops/safeDeployRewarder.ts:1]: Why `sync`?
```suggestion
import { promises as fs } from "fs";
```

File: scripts/ops/safeDeployRewarder.ts [hunk 7]
```diff
@@ -0,0 +1,91 @@
...
+  if (existsSync("./newRewarderConfig.json")) {
+    console.info("*** newRewarderConfig.json found ***");
+    try {
+      const newRewarderConfigJSONFile = readFileSync("./newRewarderConfig.json");
+      const newRewarderConfig = JSON.parse(newRewarderConfigJSONFile?.toString());
+      console.log(newRewarderConfig);
+
+      // TODO: 0xchain to add combined scripts here
+      // Add Rewarder to MCV2 etc.
+      // await proposeAddingNewRewarderToSafe(newRewarderConfig);
```
Comment (baboobhaiya) [inline scripts/ops/safeDeployRewarder.ts:72]: to add the rewarder to MCV2 we need the rewarder address
we can only get the rewarder address once we sign and send the transactions via gnosis safe

Comment (baboobhaiya) [reply scripts/ops/safeDeployRewarder.ts:72]: https://github.com/trisolaris-labs/trisolaris_core/pull/82/files#r907803852
adding LP pool to MCV2 should have multiple checks
1. LP token does not exist already in MCV2
2. owner of the rewarder contract is ops multisig

File: scripts/ops/safeDeployRewarder.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
...
+  const service = new SafeService(safeServiceURL);
+  const signer = new Wallet(safeSignerPK, provider);
+  console.log("Setup SafeEthersSigner");
+
+  const ethAdapter = new EthersAdapter({ ethers, signer });
+  const safe = await Safe.create({ ethAdapter, safeAddress });
+  const safeSigner = new SafeEthersSigner(safe, service, provider);
+
+  const complexRewarder = await ethers.getContractFactory("ComplexRewarder");
+  const rewarder = await complexRewarder.connect(safeSigner).deploy(rewardToken, lpToken, tokenPerBlock, chefV2Address);
```
[the remark this answers is not in the document]
Comment (rej156) [reply scripts/ops/safeDeployRewarder.ts:53]: do we want the owner of the Rewarder contract to be the user creating the PR?

File: scripts/ops/safeDeployRewarder.ts [hunk 9]
```diff
@@ -0,0 +1,91 @@
...
+    rewarderConfigsJSON.push(newRewarderConfig);
+
+    writeFileSync("./rewarderConfigs.json", JSON.stringify(rewarderConfigsJSON));
+    console.info("*** Added new rewarder config to rewarderConfigs.json");
+  } catch (err) {
+    console.error(err);
+  }
+};
+
+const _proposeAddingNewRewarderToSafe = async (newRewarderConfig: RewarderConfig) => {
```
[the remark this answers is not in the document]
Comment (rej156) [reply scripts/ops/safeDeployRewarder.ts:28]: unused temporarily until 0xchain updates for combined scripts to run for txs

File: scripts/ops/safeDeployRewarder.ts (hunk 8 above)
[the remark this answers is not in the document]
Comment (rej156) [reply scripts/ops/safeDeployRewarder.ts:55]: update - 0xchain said no as they're potentially not technical users

**Comment by rej156** [comment]
Work for querying pair pool with lp token address for spitting out both erc20 details to add to the config hasn't been done, do we need that for tokenlists?

**Comment by rej156** [comment]
Don't merge pls

**Pull request merged**